Sources/NIOCore/ByteBuffer-core.swift

```diff
@@ -318,12 +318,55 @@ public struct ByteBuffer {
+        // +--------------+------------------------+-------------------+
+        // |     true     | 0                      | 0                 |
+        // |     false    | 0                      | self._readerIndex |
+        // +--------------+------------------------+-------------------+
+        //
+        // The maximum value of _readerIndex (and so indexRebaseAmount) is UInt32.max, but that can only happen if the lower bound
+        // of the slice is 0, when this addition would be safe. Recall that _readerIndex is an index _into_ the slice, and the upper bound of
+        // the slice must be stored into a UInt32, so the capacity can never be more than UInt32.max. As _slice.lowerBound advances, _readerIndex
+        // must shrink. In the worst case, _readerIndex == _slice.count == (_slice.upperBound - _slice.lowerBound), so it is always safe to add
+        // _slice.lowerBound to _readerIndex. This cannot overflow.
+        //
+        // This value ends up storing the lower bound of the bytes we want to copy.
         let storageRebaseAmount = self._slice.lowerBound + indexRebaseAmount
-        let newSlice = storageRebaseAmount ..< min(storageRebaseAmount + _toCapacity(self._slice.count), self._slice.upperBound, storageRebaseAmount + capacity)
+
+        // Step 3: Here we need to find out the range within the slice that defines the upper bound of the range of bytes we want to copy.
+        // This will be the smallest of:
+        //
+        // 1. The target requested capacity, in bytes, as a UInt32 (the argument `capacity`), added to the lower bound. The resulting size of
+        //     the slice is equal to the argument `capacity`.
+        // 2. The upper bound of the current slice. This will be smaller than (1) if there are fewer bytes in the current buffer than the size of
+        //     the requested capacity.
+        //
+        // This math is checked on purpose: we should not pass a value of capacity that is larger than the size of the buffer, and if resetIndices is
+        // true then we shouldn't pass a value that is larger than readable bytes. If we do, that's an error.
+        let storageUpperBound = min(self._slice.upperBound, storageRebaseAmount + capacity)
+
+        // Step 4: Allocate the new buffer and copy the slice of bytes we want to keep.
+        let newSlice = storageRebaseAmount ..< storageUpperBound
         self._storage = self._storage.reallocSlice(newSlice, capacity: capacity)
-        self._moveReaderIndex(to: self._readerIndex - indexRebaseAmount)
-        self._moveWriterIndex(to: self._writerIndex - indexRebaseAmount)
+
+        // Step 5: Fixup the indices. These can never underflow: `indexRebaseAmount` can only ever be 0 or `self._readerIndex`, so we either do nothing
+        // or we move the indices backwards by a valid range, as `self._writerIndex` is always >= `self._readerIndex`.
+        self._moveReaderIndex(to: self._readerIndex &- indexRebaseAmount)
```

Comment:
Thanks for rewriting, commenting & fixing this! The only question I have is why we use unchecked math here. This function is the slow path and never inlined. So I don't think there's any harm caused by always doing checked math (even if it's clearly safe do use the unchecked math in _this version_ of the code.)

The reason I'm mildly worried is because of potential future changes, they might cargo cult the unchecked math even in a place where it doesn't matter and so a new vulnerability or smth may slip in.

Feel free to ignore :)

Reply:
Yeah, I'm definitely on the fence about that. Let's see what the others think.